### ISSUE TITLE
Microsoft.AspNetCore.Connections.ConnectionResetException: The client has disconnected

### ISSUE THREAD

**Issue by e-jlion** [dotnet/aspnetcore#24112]
I am using Asp.net core 3.1. Hosted in IIS, the following exceptions occasionally occur：

[日志级别]：ERROR
[日志时间]：2020-07-20 19:55:23.8917 
[日志Host]：172.16.105.122 
[日志信息]：The client has disconnected 
Microsoft.AspNetCore.Connections.ConnectionResetException: The client has disconnected
 ---> System.Runtime.InteropServices.COMException (0x80070040): 指定的网络名不再可用。 (0x80070040)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOOperation.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadBody()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadAsync(Memory`1 memory, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at System.Text.Json.JsonSerializer.ReadAsync[TValue](Stream utf8Json, Type returnType, JsonSerializerOptions options, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Formatters.SystemTextJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.Formatters.SystemTextJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
=================================

**Comment by Tratcher** [comment]
That's expected if the client cancelled the request. MVC is doing work in 5.0 to catch these exceptions.

**Comment by Tratcher** [comment]
@captainsafia please link to the MVC issue/pr that will address this.

**Comment by captainsafia** [comment]
@Tratcher Ah -- we interpreted your response as "we already fixed this as part of the 5.0 release" during triage.

If we don't, we can re-open this.

**Comment by tomerts** [comment]
i am upgrade to .net core 5.0.0-rc.2.20475.17

and I am still getting the exception in the middleware

Microsoft.AspNetCore.Connections.ConnectionResetException: The client has disconnected
 ---> System.Runtime.InteropServices.COMException (0x80070040): The specified network name is no longer available. (0x80070040)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.IIS.Core.IO.AsyncIOOperation.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadBody()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.ReadAsync(CancellationToken token)
   at System.IO.Pipelines.Pipe.DefaultPipeReader.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.IISHttpContext.ReadAsync(Memory`1 memory, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.IIS.Core.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Formatters.NewtonsoftJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value, Object container)
   at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)

I thought it would solve the problem.
How do I find the problem?

Thanks

**Comment by Tratcher** [comment]
@tomerts sorry, we didn't get this addressed in 5.0.

Note there's usually nothing wrong with the app, it's just that the client disconnected while sending the request body. The only eventual change we have planned here is to have MVC model binding catch this kind of exception and log it at a lower severity.

**Comment by tomerts** [comment]
Thanks 
1) When is a fix planned for this problem?
2) What can i do until the fix will be release so that the messages are not registered to the event viewer?

**Comment by Tratcher** [comment]
> 1. When is a fix planned for this problem?

Unknown, we'll revisit the issue as part of 6.0 planning.

> 2. What can i do until the fix will be release so that the messages are not registered to the event viewer?

You can use an exception handler middleware to handle the exception as you see fit.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/error-handling?view=aspnetcore-3.1#exception-handler-lambda-1

**Comment by deivyd321** [comment]
@Tratcher thanks for response. Is there any known steps how can it be reproduced?

Because I have some of these errors in production but app insights does not show any other info apart COMException for these requests. And I am going to add some error handling for these cases but cannot reproduce them in test and locally.

**Comment by lanfear** [comment]
I'm confused by this information a bit.

We had a fairly complex .NET Core WebApi azure hosted web-app that allowed generic file uploads from client via POST of 'file' object from client.

We had middleware in our pipeline that was catching `ConnectionResetException` to handle the case of a user disconnection (gracefully cancelling, navigating away or otherwise client-side disconnecting) during the upload file stream.

Upon updating all of our nugets to support .net core 5.x (NetCoreApp, Version=5.0" in nuget resolves, the `ConnectionResetException` stopped being hit and instead starting returning a 'graceful' (non-exception-driven) 400 'Bad Request' response.

We also updated the hosting azure web-app to specify '.NET 5 (Early Access)' in case that somehow had a side-effect on the IIS configuration running behind it causing this.

Bottom line is that this seems like a behavior change, and yet, the information above from MS seems to indicate changes were not made in the .Net Core 5.x release?

**Comment by Tratcher** [comment]
@lanfear MVC/WebApi has made incremental improvements to capture request body exceptions during the model binding step and return 400s, but only for some formatters. There should be logs to that effect.

@pranavkm @captainsafia can you summarize the MVC error handling improvements in 5.0?

**Comment by divil5000** [comment]
This is affecting us, in that it dumps an error in the Windows event viewer. It's unclear what the simplest code is, to catch and not log just that specific exception but leave the behaviour unchanged for every other exception. Can anyone help?

**Comment by captainsafia** [comment]
> @pranavkm @captainsafia can you summarize the MVC error handling improvements in 5.0?

AFAIK, we didn't do any MVC error handling improvements in 5.0....

**Comment by guidomocha** [comment]
hello!

I have this issue with a web api (asp.net core 5). the client is not cancelled any request but some different `ConnectionResetException` (with `System.Runtime.InteropServices.COMException` as inner exception) frequently is raised. it is about 2% of request. it seems there is a bug rather than cancelling the request by user!!

can anyone help us?

**Comment by captainsafia** [comment]
@guidomocha Can you open a new issue with more details about your particular scenario? Including a minimal repo with a sample app would also be helpful.

**Comment by chenjian8541** [comment]
一直有这个问题，请问该如何解决
